### ISSUE TITLE
push subjects not handled by Subscriber allow/deny SubjectPermissions

### ISSUE THREAD

**Issue by jelhub** [nats-io/nats-server#3656]
Hi,

Client using push, durable and deliverTo

Client1: subject = "tenant1.app1"
Client2: subject = "tenant2.app2"

Problem:
When using push, the subject is not handled by any allow or deny permissions set on server, however deliverTo is handled by permissions. This means, there are no way to restrict client1 or client2 to subscribe to subjects belonging to other tenants/topics.

Please advice.

Regards,
Jarle

**Comment by jelhub** [comment]
I'm using JetStream

```
sOpts.Nkeys = append(sOpts.Nkeys, &natssrv.NkeyUser{
  Nkey: subscriber.UserPublicKey,
  Permissions: &natssrv.Permissions{
    Publish: &natssrv.SubjectPermission{
      Allow: []string{"$JS.API.STREAM.NAMES", "$JS.API.CONSUMER.INFO.>", "$JS.API.CONSUMER.DURABLE.CREATE.>", "$JS.ACK.>"},
      Deny:  []string{},
	},
	Subscribe: &natssrv.SubjectPermission{
      Allow: []string{fmt.Sprintf("inbox.%s.>", tenantID), fmt.Sprintf("deliver.%s.>", tenantID)},
      Deny:  []string{},
    },
  }
}
```

The Subscribe-Allow mention above restricts "inbox" (using inboxPrefix ) and "deliver" to be tenant specific, but still the subject is not included in the allow and will pass permission validation.
even `Deny: []string{"tenant1.>", "tenant2.>"}` have no effect

delivery subject used do not prevent subject subscription e.g., "tenant1.app1"

**Comment by derekcollison** [comment]
You would need to share with us how you start your embedded server.

**Comment by jelhub** [comment]
I think permissions having allow/deny do not support push subjects

I think I instead have to use accounts/users with exports and imports?
ref: https://docs.nats.io/running-a-nats-service/configuration/securing_nats/accounts#import-configuration-map

Jarle

**Comment by derekcollison** [comment]
I am fairly confident subject permissions work as described.

See my above request for you to share more about how you are starting and configuring your embedded server.

**Comment by jelhub** [comment]
Here are startup with config-details having previously cleaned the jetstream folder and one client had connected

```
 Starting nats-server
   Version:  2.9.7
   Git:      [not set]
   Go build: go1.19.3
   Name:     NAZ4VS547CIJPRO33YCBCS643IFLEAAWY6BHPNMPJMQ7MK7IUW5Z6DIL
   Node:     mot6ayNV
   ID:       NAZ4VS547CIJPRO33YCBCS643IFLEAAWY6BHPNMPJMQ7MK7IUW5Z6DIL
 Created system account: "$SYS"
 Starting JetStream
 JetStream creating dynamic configuration - 47.79 GB memory, 1.00 TB disk
     _ ___ _____ ___ _____ ___ ___   _   __  __
  _ | | __|_   _/ __|_   _| _ \ __| /_\ |  \/  |
 | || | _|  | | \__ \ | | |   / _| / _ \| |\/| |
  \__/|___| |_| |___/ |_| |_|_\___/_/ \_\_|  |_|

          https://docs.nats.io/jetstream

 ---------------- JETSTREAM ----------------
   Max Memory:      47.79 GB
   Max Storage:     1.00 TB
   Store Directory: "jetstream"
   Encryption:      ChaCha20-Poly1305
 -------------------------------------------
   Exports:
      $JS.API.>
 Enabled JetStream for account "$G"
   Max Memory:      -1 B
   Max Storage:     -1 B
 Recovering JetStream state for account "$G"
   Stream metafile is encrypted, reading encrypted keyfile
   Starting restore for stream '$G > 6a522ccc-16a5-417b-9f62-d4fccb572738'
   Restored 0 messages for stream '$G > 6a522ccc-16a5-417b-9f62-d4fccb572738'
   Stream metafile is encrypted, reading encrypted keyfile
   Starting restore for stream '$G > 7ce89232-47e8-4eb4-8c96-6f4cc3dcb0af'
   Restored 0 messages for stream '$G > 7ce89232-47e8-4eb4-8c96-6f4cc3dcb0af'
   Recovering 1 consumers for stream - '$G > 7ce89232-47e8-4eb4-8c96-6f4cc3dcb0af'
   Consumer metafile is encrypted, reading encrypted keyfile
 JetStream state for account "$G" recovered
 Listening for client connections on 0.0.0.0:9012
 TLS required for client connections
 Get non local IPs for "0.0.0.0"
   ip=192.168.56.1
   ip=192.168.253.1
   ip=192.168.1.148
 Server is ready
```

**Comment by derekcollison** [comment]
If you are using server config we would want to see that config file to see how the permissions are being set in relation to your expectations. If you are using operator mode, we would want to see the output of `nsc describe account`. If you are embedding the server we would need to see all the setup code wrt permissions and how you are starting the server.

**Comment by jelhub** [comment]
Don't use nats config file

Code below shows startup details/config

```
fs := flag.NewFlagSet("streaming", flag.ExitOnError)
sOpts, err := natssrv.ConfigureOptions(
  fs,
  nil,
  func() {
    fmt.Printf("nats-server version %s, ", natssrv.VERSION)
    natssrv.PrintServerAndExit()
  },
  fs.Usage,
  natssrv.PrintTLSHelpAndDie)

if err != nil {
  natssrv.PrintAndDie(err.Error())
}

sOpts.JetStream = true
sOpts.StoreDir = ExPath
sOpts.Host = "0.0.0.0"
sOpts.Port = Config.StreamServer.Port
if Config.StreamServer.EncryptionAtRestKey != "" {
  sOpts.JetStreamKey = Config.StreamServer.EncryptionAtRestKey
}

// permissions - publisher
sOpts.Nkeys = []*natssrv.NkeyUser{}
sOpts.Nkeys = append(sOpts.Nkeys, &natssrv.NkeyUser{
  Nkey: Config.StreamServer.Publisher.UserPublicKey,
  Permissions: &natssrv.Permissions{
    Publish: &natssrv.SubjectPermission{
      Allow: []string{">"},
      Deny:  []string{},
    },
    Subscribe: &natssrv.SubjectPermission{
      Allow: []string{">"},
      Deny:  []string{},
    },
  },
})

// permissions - subscribers
for _, tenant := range Config.Tenants {
  for _, subscriber := range tenant.General.StreamSubscribers {
    sOpts.Nkeys = append(sOpts.Nkeys, &natssrv.NkeyUser{
      Nkey: subscriber.UserPublicKey,
      Permissions: &natssrv.Permissions{
        Publish: &natssrv.SubjectPermission{
          Allow: []string{"$JS.API.STREAM.NAMES", "$JS.API.CONSUMER.INFO.>", "$JS.API.CONSUMER.DURABLE.CREATE.>", "$JS.ACK.>"},
          Deny:  []string{},
        },
        Subscribe: &natssrv.SubjectPermission{
          //Allow: []string{"_INBOX.>", fmt.Sprintf("deliver.%s.>", tenant.General.TenantID)},
          Allow: []string{fmt.Sprintf("inbox.%s.>", tenant.General.TenantID), fmt.Sprintf("deliver.%s.>", tenant.General.TenantID)},
          Deny:  []string{},
        },
      },
    })
  }
}

...
some tls config
...
sOpts.TLSConfig = tlsConfig

if err := natssrv.Run(nSrv); err != nil {
  log.Fatal().Msgf("nats server startup error: %s", err.Error())
}
nSrv.WaitForShutdown()
```

**Comment by derekcollison** [comment]
And what specifically is not working? What is the expectation?

Also for publishing, that is describing the default, and is not necessary. By default we allow everything and deny nothing.

**Comment by ripienaar** [comment]
Are you saying that user1 can make a consumer with filter subject set to user2 subject? 

Hard to follow exactly what your problem is here.

**Comment by jelhub** [comment]
As you see permissions for subscribers do not include any allow/deny for subjects beeing used (only mandatory inbox and deliver), but still clients can subscribe to any subject they want.

In subscriber allow I would like to include rule for subjects like: `fmt.Sprintf("%s.>", tenant.General.TenantID)`
e.g.: "tenant1.>" which should cover subscriber/consumer subject, but doing so have no effect (nor any deny rules)

Having two subscribers from different tenants:
Client1 (tenant1): subject = "tenant1.app1"
Client2 (tenant2): subject = "tenant2.app2"

They should only be allowed to subscribe to the "tentant" subject they belongs to.
Now, they can choose whatever they want (because nats routes subjects to deliverTo when using push and seems not to evaluate any subject permissions defined)

**Comment by jelhub** [comment]
> Also for publishing, that is describing the default, and is not necessary. By default we allow everything and deny nothing.

Publisher is protected by nkey auth and needs to be included even using default allow/deny

**Comment by ripienaar** [comment]
If you are using at least server 2.9.0 and go client 1.17.0 you can then add additional ACLs on the consumer create to limit this.

You can deny access to `$JS.API.CONSUMER.CREATE.<STREAM>` and `$JS.API.CONSUMER.DURABLE.CREATE.<STREAM>.*` and allow access to `$JS.API.CONSUMER.CREATE.<STREAM>.*.tenant1.>` this should then prevent them from adding consumers on other filter subjects and prvent older clients from using the old subjects

**Comment by jelhub** [comment]
Startup trace given above shows server Version:  2.9.7
I'm using nodejs module "nats": "^2.7.1"

**Comment by jelhub** [comment]
Stream names are set to tenantid
Doing some redesign and not including tenantid in subject

publisher#2 adding stream name=tentant2 fails on subject="app.>" because publisher#1 stream name=tenant1 already use subject="app.>"

Error message:
error adding stream: nats: subjects overlap with an existing stream

Seems to be something like:
https://giters.com/nats-io/nats-server/issues/2144

Are multiple streams using same subject supported?

**Comment by derekcollison** [comment]
Not within same account, a single subject within an account needs to be processed by a single stream.

**Comment by jelhub** [comment]
Now tested publisher#1 and publisher#2 using different Nkey (they will then not the same account), but still the same `subjects overlap with an existing stream` error as mention above.

**Comment by derekcollison** [comment]
Explain what you are trying to accomplish and want and we will set up a system config you can copy.

**Comment by jelhub** [comment]
Well, you could then focus on the main problem:
Setup a config that shows `push subjects are handled by permissions` 

This issue is about multi-tenancy, clients at one tenant should not be able to subscribe to other tenants

Using JetStream, push, durable and deliverTo

Having two publishers using different tenants:
Publisher#1: stream=tenant1, subject="tenant1.>"
Publisher#2: stream=tenant2, subject="tenant2.>"
(subject="app.>" does not work as mention above if both publishers using same - second publisher startup will then fail)

Having two subscribers using different tenants:
Client1: subject="tenant1.app1"
Client2: subject="tenant2.app2"

All protected by Nkey and Allow/Deny permissions for Publish/Subscribe
**Client1 (tenant1) should only be allowed subscribing to subject "tenant1.>"
Client2 (tenant2) should only be allowed subscribing to subject "tenant2.>"**

Problem:
Changing Client2: subject="tenant1.app1"
(both clients using same subject/tenant)
I'm not able to deny clients subscribing to subjects (tentants) they are not allowed to use.
Subject is not evaluated by permission policies, even Deny="tenant1.>, tenant2.>" have no effect.

Thanks,
Jarle

**Comment by ripienaar** [comment]
When you do js.Subscribe() it isn’t a subscribe like you might think - this API makes you think it does sometning when it doesn’t. 

What it actually does is set up a consumer with a filter subject. And you need to set up the permissions as I described.

**Comment by jelhub** [comment]
Closing summary:

Using push, durable and deliverTo

`push subjects are not handled by Subscriber allow/deny SubjectPermissions`
however:
`push subjects are handled by Publisher allow/deny SubjectPermissions`

e.g.:
Client1 subscribing subject=tenant1.app1
Client2 subscribing subject=tenant2.app2

Following **Subscriber** rule will protect cross tenant subscription:
```
for _, subscriber := range subscribers {
  sOpts.Nkeys = append(sOpts.Nkeys, &natssrv.NkeyUser{
    Nkey: subscriber.UserPublicKey,
    Permissions: &natssrv.Permissions{
      Publish: &natssrv.SubjectPermission{
        // allow rule preventing cross tenant subscription
        Allow: []string{"$JS.API.STREAM.NAMES", fmt.Sprintf("$JS.API.CONSUMER.INFO.%s.>", subscriber.tenantId), fmt.Sprintf("$JS.API.CONSUMER.CREATE.%s.>", subscriber.tenantId), fmt.Sprintf("$JS.API.CONSUMER.DURABLE.CREATE.%s.*", subscriber.tenantId)},
        Deny:  []string{},
      },
      Subscribe: &natssrv.SubjectPermission{
         // below not mandatory for preventing cross tenant subscription, but for other security reasons
         // subscriber using inboxPrefix "inbox.tenant" instead of default "_INBOX" and deliverTo prefix "deliver.tenant"
        Allow: []string{fmt.Sprintf("inbox.%s.>", subscriber.tenantId), fmt.Sprintf("deliver.%s.>", subscriber.tenantId)},
        Deny:  []string{},
      },
    },
  })
}
```

Thanks,
Jarle
